Sources/Basics/FileSystem+Extensions.swift

```diff
@@ -17,13 +20,93 @@ extension FileSystem {
+        self.cachesDirectory
+    }
+
     /// SwiftPM cache directory under user's caches directory (if exists)
     public var swiftPMCacheDirectory: AbsolutePath {
-        if let cachesDirectory = self.cachesDirectory {
-            return cachesDirectory.appending(component: "org.swift.swiftpm")
+        if let path = self.idiomaticUserCacheDirectory {
+            return path.appending(component: "org.swift.swiftpm")
         } else {
-            return self.dotSwiftPM.appending(component: "cache")
+            return self.dotSwiftPMCachesDirectory
+        }
+    }
+
+    fileprivate var dotSwiftPMCachesDirectory: AbsolutePath {
+        return self.dotSwiftPM.appending(component: "cache")
+    }
+}
+
+extension FileSystem {
+    public func getOrCreateSwiftPMCacheDirectory() throws -> AbsolutePath {
+        let idiomaticCacheDirectory = self.swiftPMCacheDirectory
+        // Create idiomatic if necessary
+        if !self.exists(idiomaticCacheDirectory) {
+            try self.createDirectory(idiomaticCacheDirectory, recursive: true)
+        }
+        // Create ~/.swiftpm if necessary
+        if !self.exists(self.dotSwiftPM) {
+            try self.createDirectory(self.dotSwiftPM, recursive: true)
+        }
+        // Create ~/.swiftpm/cache symlink if necessary
+        if !self.exists(self.dotSwiftPMCachesDirectory, followSymlink: false) {
+            try self.createSymbolicLink(dotSwiftPMCachesDirectory, pointingAt: idiomaticCacheDirectory, relative: false)
+        }
+        return idiomaticCacheDirectory
+    }
+}
+
+// MARK: - config
+
+extension FileSystem {
+    private var idiomaticUserConfigDirectory: AbsolutePath? {
+        return FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first.flatMap { AbsolutePath($0.path) }
+    }
+
+    /// SwiftPM config directory under user's config directory (if exists)
+    public var swiftPMConfigDirectory: AbsolutePath {
+        if let path = self.idiomaticUserConfigDirectory {
+            return path.appending(component: "org.swift.swiftpm")
+        } else {
+            return self.dotSwiftPMConfigDirectory
+        }
+    }
+
+    fileprivate var dotSwiftPMConfigDirectory: AbsolutePath {
+        return self.dotSwiftPM.appending(component: "config")
+    }
+}
+
+extension FileSystem {
+    public func getOrCreateSwiftPMConfigDirectory() throws -> AbsolutePath {
+        let idiomaticConfigDirectory = self.swiftPMConfigDirectory
+
+        // temporary 5.5, remove on next version: transition from ~/.swiftpm/config to idiomatic location + symbolic link
+        if idiomaticConfigDirectory != self.dotSwiftPMConfigDirectory &&
+            self.exists(self.dotSwiftPMConfigDirectory) && self.isDirectory(self.dotSwiftPMConfigDirectory) &&
+            !self.exists(idiomaticConfigDirectory) {
+            print("transitioning \(self.dotSwiftPMConfigDirectory) to \(idiomaticConfigDirectory)")
```

Comment:
With this code now being down in a shared module, does it make sense to have the print statement here directly, or should that be through a callback?  Just a thought — possibly this code is specific enough that it's known to not be invoked from IDEs or other clients of libSwiftPM?

Reply:
the idea here was to print debugging info in case someone runs into issues while transitioning between versions of 5.5. this check (L91-L95) should be removed in a few weeks

Reply:
do you have another idea, happy to change!

Reply:
No, missed the part that it won't be there in the long run.  Thanks!